Feed retrieved properly from network using Arrow

raulh82vlc opened the pull request:
From Network datasource through the Repository to the UseCase, then to the Presentation using Arrow

raulh82vlc committed:
Feed retrieved from Network datasource through the Repository to the UseCase to the Presentation using Arrow

raulh82vlc committed:
Fixed visibility when showResults for the List

raulh82vlc commented on app/src/main/java/com/raulh82vlc/functionalkotlinarrow/data/network/NetworkDataSource.kt:
```diff
@@ -0,0 +1,134 @@
+import arrow.effects.typeclasses.Async
+import arrow.typeclasses.binding
+import com.google.gson.Gson
+import com.jakewharton.retrofit2.adapter.kotlin.coroutines.experimental.CoroutineCallAdapterFactory
```
I was trying here to get a Deferred and play with it before heading to a normal _Call_ object

suhtai commented on app/src/main/java/com/raulh82vlc/functionalkotlinarrow/FunctionalKotlinApp.java:
```diff
@@ -0,0 +1,44 @@
+import android.app.Application;
+
+
+import timber.log.Timber;
```
we could make the imports go together

raulh82vlc replied:
done

suhtai commented on app/src/main/java/com/raulh82vlc/functionalkotlinarrow/FunctionalKotlinApp.java:
```diff
@@ -0,0 +1,44 @@
+        if (BuildConfig.DEBUG) {
+            Timber.plant(new Timber.DebugTree());
+            Timber.tag("KotlinFunctional");
+            Timber.d("KotlinFunctional -> onCreate");
```
These can be collapsed to be `Timber.tag("TAG").d("MESSAGE")`

raulh82vlc replied:
done

suhtai commented on app/src/main/java/com/raulh82vlc/functionalkotlinarrow/FunctionalKotlinApp.java:
```diff
@@ -0,0 +1,44 @@
+ *
+ * @author Raul Hernandez Lopez
+ */
+public class FunctionalKotlinApp extends Application {
```
Should this be a Kotlin class?

raulh82vlc replied:
👍

suhtai commented on app/src/main/java/com/raulh82vlc/functionalkotlinarrow/data/FeedRepository.kt:
```diff
@@ -0,0 +1,27 @@
+import arrow.effects.IO
+import com.raulh82vlc.functionalkotlinarrow.data.cache.model.FeedItemCacheModel
+import com.raulh82vlc.functionalkotlinarrow.data.network.NetworkDataSource
+object FeedRepository {
```
extra line needed for readability

raulh82vlc replied:
done

suhtai commented on app/src/main/java/com/raulh82vlc/functionalkotlinarrow/data/NetworkToCacheMapper.kt:
```diff
@@ -0,0 +1,53 @@
+            )
+        }
+        return cacheOutput
+    }
```
If this function just does the conversion from a `FeedItemApiModel` to `FeedItemCacheModel` then when using this on a list you can do it directly on the list itself:
The function would look like:
```
fun FeedItemApiModel.toFeedItemCacheModel : FeedItemCacheModel = 
  FeedItemCacheModel(
    title,
    link,
    FeedItemCacheModel.MediaItem(media.url),
    description,
    dateTaken,
    published,
    author,
    authorId,
    mapTags(apiItem.tags)
  )
```
You can also add name parameters to make it more clear (and when changing the order of the params this doesn't require changes)
The usage would be:
```
listOfApiModel.map { it.toFeedItemCacheModel() }
```

raulh82vlc replied:
I don´t like to mix different layer models into other ones, that's why mappers serve well to decouple code, to avoid mixing models

suhtai replied:
But we are also mixing them on the mapper function. Think of it as a first-class function in this case, or a tiny class/type :)
You mapper becomes the function (This sounds more sentai than I intended XD)

raulh82vlc replied:
I'll keep in mind, thanks a lot for this alternative, but I'll keep the mapper at this time

suhtai commented on app/src/main/java/com/raulh82vlc/functionalkotlinarrow/data/NetworkToCacheMapper.kt:
```diff
@@ -0,0 +1,53 @@
+        return cacheOutput
+    }
+
+    fun mapTags(tags: String): List<String> {
```
This can be done like this:
```
fun mapTags(tags: String): List<String> = 
    tags.split(REGEX_WHITESPACE.toRegex())
        .map { it.trim() }
        .filter { it.isNotEmpty() }
```
With less overhead. Also, if it's only used here, it probably should be private. And once it's private then you can make it an extension function of `String` and call it something like `toTagList`

raulh82vlc replied:
it could be private, but this was not private at the java project I translated it from, because the unit tests that are not here yet

suhtai replied:
The tests shouldn't case for this function really, as it can be tested with the mapper function. this way you can change the implementation easier in the future without having to rewrite the test

suhtai commented on app/src/main/java/com/raulh82vlc/functionalkotlinarrow/data/cache/CacheDataSourceImpl.kt:
```diff
@@ -0,0 +1,49 @@
+
+    override fun isEmpty() = listOfItems.isEmpty()
+
+    override fun getItemFromFeed(title: String, authorId: String): FeedItemCacheModel?
```
Better to return `Option` instead of a `nullable` as it allows you to transform between different applicative types.

raulh82vlc replied:
yes, using optionals is a very good idea, good point, this is not used yet, will be for a next PR though

raulh82vlc replied:
done

suhtai commented on app/src/main/java/com/raulh82vlc/functionalkotlinarrow/data/cache/CacheDataSourceImpl.kt:
```diff
@@ -0,0 +1,49 @@
+            listOfItems.add(item)
+            dictionaryOfItems[item.title + item.authorId] = item
+        }
+    }
```
If you make both `listOfItems` and `dictionaryOfItems` writable (using `var` instead of `val`) you can define them like:
```
private var listOfItems = emptyList<FeedItemCacheModel>()
private var dictionaryOfItems = emptyMap<String, FeedItemCacheModel>()
```
And then, when you save the feed you can do:
```
listOfItems = listOf(itemCacheModels)
dictionaryOfItems = listOfItems.map { "${it.title}${it.authorId}" to it }.toMap()
```
This makes it thread save and it's only two lines.

raulh82vlc replied:
being both of the cases mutable, in the case I have because the elements are, and in yours because you change a reference to a new collection. is it safer because the final result is passed as the latest instance? rather than changing value per value in several iterations?

it would be very difficult to have a problem anyways since it would be one request unless there is a rotation (something that probably will be prevented with a better cache policy in a next future)

suhtai replied:
Well, if you are iterating over a list and someone modifies it unless you define it sync it'll blow up. By replacing the list (not mutating) then if someone is already iterating over it there would be no problem and we just assign the new list. There are also problems with list sizes. As if the new list is bigger the `ArrayList` will have to increment the internal array... but that's not that important as you aren't gonna have large amounts of data there.

raulh82vlc replied:
done! thank you

suhtai commented on app/src/main/java/com/raulh82vlc/functionalkotlinarrow/data/cache/model/FeedItemCacheModel.kt:
```diff
@@ -0,0 +1,33 @@
+ * FeedItem for the Cache layer, which is the data common model
+ * @author Raul Hernandez Lopez.
+ */
+data class FeedItemCacheModel(val title: String,
```
This would read better if they are in a new line:
```
data class SomeType(
  val paramOne: String,
  val paramTwo: String
)
```

suhtai replied:
Also, it seems redundant to have the postfix `Model` I would have called this `CachedFeedItem`

raulh82vlc replied:
matter of preference

suhtai commented on app/src/main/java/com/raulh82vlc/functionalkotlinarrow/data/network/NetworkDataSource.kt:
```diff
@@ -0,0 +1,134 @@
+        return emptyList()
+    }
+
+    private fun hasFeedFormat(responseBodyResult: String): Boolean =
```
This can have a `String` as the receiver so can do `string.hasFeedFormat()`

raulh82vlc replied:
like an extension function you mean?

suhtai replied:
Yeah, sorry the receiver in an extension function is the type you are extending 😃

raulh82vlc replied:
done!

suhtai commented on app/src/main/java/com/raulh82vlc/functionalkotlinarrow/data/network/NetworkDataSource.kt:
```diff
@@ -0,0 +1,134 @@
+            return emptyList()
+        }
+        return emptyList()
+    }
```
```
private fun transformToRepository(feed: Call<ResponseBody>) =
  execute(feed)?.string()?.takeIf { it.hasFeedFormat() }?.let { response ->
    NetworkToCacheMapper.map(deserializeFeedJSON(extractJSONFromResponse(response)).feedItems)
  } ?: emptyList()
```

raulh82vlc replied:
good one

suhtai replied:
Now it hit me that we have Arrow available. It may be easier to read with an `Option`:
```
execute(feed).toOption()
  .map { it.string() }
  .filter { it.hasFeedFormat() }
  .map { it.extractJSONFromResponse().toFeedItemCacheModel().feedItems) }
  .getOrElse { emptyList() }
```

suhtai commented on app/src/main/java/com/raulh82vlc/functionalkotlinarrow/data/network/NetworkDataSource.kt:
```diff
@@ -0,0 +1,134 @@
+
+    fun requestFeed(format: String): IO<List<FeedItemCacheModel>> {
+        val monadError = IO.monadError()
+        return monadError.binding {
```
This could be inlined and have an expression body (without the curly brackets)

suhtai commented on app/src/main/java/com/raulh82vlc/functionalkotlinarrow/data/network/NetworkDataSource.kt:
```diff
@@ -0,0 +1,134 @@
+                    onSuccess = { monadError.just(it) },
+                    AC = IO.async()
+            ).bind()
+            result.bind()
```
This could be inlined and do `bind().bind()` in the end instead.

raulh82vlc replied:
it could, but looking at this, which was like this at @JorgeCastilloPrz sample project, I think this is more clear, because **bind().bind()** doesn´t feel very intuitive about what this does

suhtai replied:
```
fun MonadError.superbind() = bind().bind()
```
:trollface:

raulh82vlc replied:
done!

suhtai commented on app/src/main/java/com/raulh82vlc/functionalkotlinarrow/data/network/NetworkDataSource.kt:
```diff
@@ -0,0 +1,134 @@
+    private fun extractJSONFromResponse(response: String): String {
+        val responseOutput = response.replaceFirst(JSON_FLICKR_FEED_KEYWORD.toRegex(), "")
+        return responseOutput.substring(0, responseOutput.length - 1)
+    }
```
With receiver:
```
private fun String.extractJSONFromResponse(): String =
  replaceFirst(JSON_FLICKR_FEED_KEYWORD.toRegex(), "")
```
Also, that last `substring` seems a bit redundant

raulh82vlc replied:
first is the answer to the prev questions, cheers!
the second I can´t remember when I did it in java if it was necessary for some reason, I´ll check it out during the weekend

raulh82vlc replied:
this was required to remove the latest ) from the malformed json that the flickr API is given us, checked a couple of days ago when I did some changes according to your suggestions

suhtai commented on app/src/main/java/com/raulh82vlc/functionalkotlinarrow/data/network/NetworkDataSource.kt:
```diff
@@ -0,0 +1,134 @@
+    }
+
+    @Throws(ApiException::class)
+    fun execute(call: Call<ResponseBody>): ResponseBody? {
```
Instead of returning a `nullable` in here you can return `Try<ResponseBody>` and it doesn't need the `@Throws` annotation.
```
fun Call<ResponseBody>.tryToExecute(): Try<ResponseBody> = 
  Try { execute() }
    .recoverWith(Try.raise(ApiException("Network error", e)))
    .flatMap { response ->
      response?.takeIf { it.isSuccessful }?.body() ?: throw ApiException(response.errorBody().toString(), null)
    }
```

raulh82vlc replied:
I have to learn a lot about this using Arrow, cheers for this

raulh82vlc replied:
will do for a new refactor, not for this one

suhtai commented on app/src/main/java/com/raulh82vlc/functionalkotlinarrow/data/network/NetworkDataSource.kt:
```diff
@@ -0,0 +1,134 @@
+                .connectTimeout(3, TimeUnit.SECONDS)
+                .writeTimeout(10, TimeUnit.SECONDS)
+                .readTimeout(10, TimeUnit.SECONDS)
+        return okHttpClient.build()
```
I would avoid using a temporary property here and use the function chaning with an expression body. In fact, given that this is only used here, I would just make it a property instead of a function. (potentially private too)

suhtai commented on app/src/main/java/com/raulh82vlc/functionalkotlinarrow/data/network/error/ApiException.kt:
```diff
@@ -0,0 +1,19 @@
+
+package com.raulh82vlc.functionalkotlinarrow.data.network.error
+
+class ApiException(msg: String, cause: Throwable?) : Exception(msg, cause)
```
Rather than use exceptions, it may be better to have a closed hierarchy of error types (seal class) with the possible errors. Also means that we don't have to pass around Exceptions.

raulh82vlc replied:
This is a total refactor though, I'll keep it in mind for the next round then

suhtai commented on app/src/main/java/com/raulh82vlc/functionalkotlinarrow/presentation/Navigation.kt:
```diff
@@ -0,0 +1,34 @@
+//        val intent = Intent(ctx, FeedDetailActivity::class.java)
+//        intent.putExtra(
+//                EXTRA_FEED_ID, feedId)
+//        ctx.startActivity(intent)
```
Is this code intended to be commented out?

raulh82vlc replied:
it is, there is no detail yet

suhtai commented on app/src/main/java/com/raulh82vlc/functionalkotlinarrow/presentation/Navigation.kt:
```diff
@@ -0,0 +1,34 @@
+    fun goToFeedDetail(ctx: Context, feedId: String): IO<Unit> =
+            IO.applicative().just(launch(ctx, feedId)).fix()
+
+    private fun launch(ctx: Context, feedId: String) {
```
I would personally avoid using `ctx` and just have the full name of the type.

suhtai commented on app/src/main/java/com/raulh82vlc/functionalkotlinarrow/presentation/Presentation.kt:
```diff
@@ -0,0 +1,49 @@
+
+            monadError.just(view.showFeed(result)).bind()
+        }.fix()
+    }
```
This seems odd, I'll have to have a more detailed look at it...

raulh82vlc replied:
refactored

suhtai commented on app/src/main/java/com/raulh82vlc/functionalkotlinarrow/ui/MainActivity.kt:
```diff
@@ -16,14 +16,60 @@
 
-class MainActivity : AppCompatActivity() {
+class MainActivity : AppCompatActivity(), FeedListView {
+    private lateinit var feedAdapter: FeedAdapter
```
Humongous NO! Don't use `lateinit` if you can avoid it. This can be done more easily with `by lazy {  }`.

raulh82vlc replied:
honestly, I need to read more about differences between both, I got used to do this with **lateinit** all the time, why does it have a problem?

suhtai replied:
Actually, in this case, since the adapter is actually not dependent on the activity it could just be instantiated with the property declaration instead of in `onCreate`. Same as in Java 😄

raulh82vlc replied:
done

suhtai commented on app/src/main/java/com/raulh82vlc/functionalkotlinarrow/ui/MainActivity.kt:
```diff
@@ -16,14 +16,60 @@
     override fun onCreate(savedInstanceState: Bundle?) {
         super.onCreate(savedInstanceState)
         setContentView(R.layout.activity_main)
+        setFeedList()
```
Named like a setter but doesn't have an argument. Makes it confusing. I would call it either `setupFeedList` or `initFeedList`

raulh82vlc replied:
yep, agreed

suhtai commented on app/src/main/java/com/raulh82vlc/functionalkotlinarrow/ui/MainActivity.kt:
```diff
@@ -16,14 +16,60 @@
+    private fun showResults() {
+        no_results_view.visibility = View.GONE
+        feed_rv.visibility = View.VISIBLE
     }
```
This can be modelled with a sealed class `State` type:
```
sealed class State {
  object Empty : State()
  data class Failed(val message: String) : State()
  data class Ready(val feed: List<FeedItemCacheModel>) : State()
}
```

raulh82vlc replied:
good one, seems legit

raulh82vlc replied:
will do for a new refactor though

suhtai commented on app/src/main/java/com/raulh82vlc/functionalkotlinarrow/ui/adapter/FeedAdapter.kt:
```diff
@@ -0,0 +1,60 @@
+
+class FeedAdapter(
+        var feedList: List<FeedItemCacheModel> = ArrayList(),
+        val itemClick: (FeedItemCacheModel) -> Unit) : RecyclerView.Adapter<FeedAdapter.FeedViewHolder>() {
```
When having multiline parameters, I personally find it easier to read if the extension of the class is made on a new line.

suhtai commented on app/src/main/java/com/raulh82vlc/functionalkotlinarrow/ui/adapter/FeedAdapter.kt:
```diff
@@ -0,0 +1,60 @@
+
+    class FeedViewHolder(view: View,
+                     val itemClick: (FeedItemCacheModel) -> Unit) : RecyclerView.ViewHolder(
+            view) {
```
Odd choice of indentation :trollface:

suhtai commented on app/src/main/java/com/raulh82vlc/functionalkotlinarrow/ui/adapter/FeedAdapter.kt:
```diff
@@ -0,0 +1,60 @@
+
+        fun bind(itemFeed: FeedItemCacheModel) {
+            with(itemFeed) {
+                com.squareup.picasso.Picasso.with(itemView.context)
```
You can import and can rename it as `withPicasso`:
```
import com.squareup.picasso.Picasso.with as withPicasso
```

raulh82vlc replied:
good one!

suhtai commented on app/src/main/java/com/raulh82vlc/functionalkotlinarrow/ui/adapter/FeedAdapter.kt:
```diff
@@ -0,0 +1,60 @@
+            with(itemFeed) {
+                com.squareup.picasso.Picasso.with(itemView.context)
+                        .load(itemFeed.media.url).into(itemView.iv_image)
+                itemView.txt_title.text = itemFeed.title
```
If you are doing `with(itemFeed)` then there is no need to reference `itemFeed` here again (or at all inside this scope)

raulh82vlc replied:
absolutely, I only had a first scan to this file I believe, good catch!

suhtai commented on app/src/main/java/com/raulh82vlc/functionalkotlinarrow/ui/adapter/FeedAdapter.kt:
```diff
@@ -0,0 +1,60 @@
+        notifyDataSetChanged()
+    }
+
+    class FeedViewHolder(view: View,
```
I would suggest implementing `LayoutContainer` here. Otherwise, each time `bind` is called it does a new lookup of the views... 
https://stackoverflow.com/questions/45246268/do-kotlin-android-extensions-cache-the-synthetic-properties-or-each-time-it-call

raulh82vlc replied:
perfect! 👍 this way https://github.com/antoniolg/Kotlin-for-Android-Developers/blob/master/app/src/main/java/com/antonioleiva/weatherapp/ui/adapters/ForecastListAdapter.kt

suhtai commented on app/src/main/res/layout/item_list.xml:
```diff
@@ -0,0 +1,54 @@
+  ~ limitations under the License.
+  -->
+
+<LinearLayout
```
`ConstraintsLayout`? :D

raulh82vlc replied:
for a complex view with good animations for sure I'd go for it, for only a tuple into a simple adapter seems to be a little overhead to me (matter of personal preference of course). I'd use a _CoordinatorLayout'em_ all approach, at a more complex and nicer (visually speaking) side project, this is purely for trying to learn a little bit of functional really

suhtai commented on app/src/main/res/values/styles.xml:
```diff
@@ -1,11 +1,37 @@
+        <item name="android:text">@string/list_empty</item>
+        <item name="android:background">@android:color/white</item>
+    </style>
+
```
Many of the properties here seem like they are only used in one place, extracting these may be early optimisation. Styles are good for reusability... Setting `android:text` in a style is too much indirection, and unforeseen effects under the hood.

raulh82vlc replied:
it's a side project, so this has inherited many things I had before, for this doesn't really change a lot. but I see your point

raulh82vlc commented on app/src/main/java/com/raulh82vlc/functionalkotlinarrow/data/FeedRepository.kt:
```diff
@@ -0,0 +1,27 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
```
once a license is at the project associated, every file you create adds it, so...

pablisco replied:
You can edit the template ;)

raulh82vlc committed:
Added suggestions
- Reusing viewholder with cached one: LayoutContainer for caching itemView at Adapter
- Enabled androidExtensions for it
- transform App file to kt

Other Optimisations
- Make attributes from NetworkDataSource private

suhtai commented on app/src/main/java/com/raulh82vlc/functionalkotlinarrow/ui/adapter/FeedAdapter.kt:
```diff
@@ -44,15 +46,15 @@ class FeedAdapter(
-            view) {
+    class FeedViewHolder(override val containerView: View?,
+                         val itemClick: (FeedItemCacheModel) -> Unit) : RecyclerView.ViewHolder(
+            containerView),
```
This indentation hurts my eyes 😛

suhtai replied:
I would do this instead:
```
class FeedViewHolder(
  override val containerView: View,
  val itemClick: (FeedItemCacheModel) -> Unit
) : ViewHolder(containerView), LayoutContainer {
```

suhtai commented on app/src/main/java/com/raulh82vlc/functionalkotlinarrow/data/NetworkToCacheMapper.kt:
```diff
@@ -0,0 +1,53 @@
+import com.raulh82vlc.functionalkotlinarrow.data.network.model.FeedItemApiModel
+import java.util.*
+
+object NetworkToCacheMapper {
```
This object is not really needed if the map function is defined as an extension function

raulh82vlc replied:
good point, I'll apply it

suhtai commented on app/src/main/java/com/raulh82vlc/functionalkotlinarrow/data/cache/CacheDataSourceImpl.kt:
```diff
@@ -0,0 +1,49 @@
+ *
+ * @author Raul Hernandez Lopez
+ */
+object CacheDataSourceImpl : CacheDataSource {
```
Using the `Impl` indicates that probably the interface is not needed

raulh82vlc replied:
probably a interface doesn´t do a lot for this, needs an object to keep all from the object as a singleton, that´s for sure. nothing major for this

suhtai commented on app/src/main/java/com/raulh82vlc/functionalkotlinarrow/data/network/NetworkDataSource.kt:
```diff
@@ -0,0 +1,136 @@
+/**
+ * Data source responsible of network requests
+ */
+object NetworkDataSource {
```
Same as before. If the function is defined as an extension function this object may not be needed

raulh82vlc replied:
I'd prefer to keep it as an object, to maintain in memory while the app is alive, all different values created on it, like gson and retrofit, this really needs a singleton

suhtai commented on app/src/main/java/com/raulh82vlc/functionalkotlinarrow/presentation/Navigation.kt:
```diff
@@ -0,0 +1,34 @@
+import arrow.effects.applicative
+import arrow.effects.fix
+
+object Navigation {
```
Another object that can be removed if functions are defined as extensions

raulh82vlc replied:
good candidate for extension functions, true for this one

suhtai commented on app/src/main/java/com/raulh82vlc/functionalkotlinarrow/presentation/Presentation.kt:
```diff
@@ -0,0 +1,49 @@
+    fun showFeed(feed: List<FeedItemCacheModel>)
+}
+
+object Presentation {
```
This object could also be avoided

raulh82vlc replied:
done

raulh82vlc committed:
Suggestions amended
- NetworkToCacheMapper now is a extension function
- transformToRepository method at NetworkDataSource uses all the power of Arrow

raulh82vlc reviewed (commented):
all ready to merge @pablisco
